subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -4197,3 +4310,166 @@ int bt_hfp_ag_inband_ringtone(struct bt_hfp_ag *ag, bool inband)
 
 	return err;
 }
+
+int bt_hfp_ag_voice_recognition(struct bt_hfp_ag *ag, bool activate)
+{
+#if defined(CONFIG_BT_HFP_AG_VOICE_RECG)
+	int err;
+	bool feature;
+	char *bvra;
+
+	LOG_DBG("");
+
+	if (ag == NULL) {
+		return -EINVAL;
+	}
+
+	hfp_ag_lock(ag);
+	if (ag->state != BT_HFP_CONNECTED) {
+		hfp_ag_unlock(ag);
+		return -ENOTCONN;
+	}
+	hfp_ag_unlock(ag);
+
+	if (activate && atomic_test_bit(ag->flags, BT_HFP_AG_VRE_ACTIVATE)) {
+		LOG_WRN("VR has been activated");
+		return -ENOTSUP;
+	} else if (!activate && !atomic_test_bit(ag->flags, BT_HFP_AG_VRE_ACTIVATE)) {
+		LOG_WRN("VR is not activated");
+		return -ENOTSUP;
+	}
+
+	hfp_ag_lock(ag);
+	feature = (ag->ag_features & BT_HFP_AG_FEATURE_ENH_VOICE_RECG) &&
+		(ag->hf_features & BT_HFP_HF_FEATURE_ENH_VOICE_RECG);
+	hfp_ag_unlock(ag);
```

Comment:
This really looks like too granular locking to me, which in turn unnecessarily increases the code size and hurts readability. Unless you've actually profiled it and have good reasons not to do it, I'd recommend to avoid multiple critical sections for the same shared object within a single function.

Reply:
Remove line 4342 and 4345